# <code style="background:yellow;color:black">Summary of today's session:</code> 

1. Data slicing:
    * Data slicing through iloc
    * Data slicing through loc
    * Data slicing by extracting one column
    * Data slicing by extracting multiple columns
    * Data slicing by combining slicing and multi-indexing
    
2. In slicing, we can also provide step size or jump size, both in iloc and loc

3. Addition of new rows:
    * append() method
    * loc Indexer
    
4. Modification of existing rows:
    * Modifying an existing row using loc
    * Modifying or updating an existing row using iloc
    
5. Duplicate rows:
    * Checking for duplicate rows
    * Storing of duplicate rows
    * Deleting Duplicate Rows

6. Deletion of rows and columns:
    * drop() method
    * Deleting a single row
    * Deleting multiple rows
    * Deleting a single column
    * Deleting multiple columns
    * Deleting rows and columns permanently

7. Resetting index: reset_index() method

8. Aggregate functions:
    * Sum (sum()): Calculates the sum of values in a group.
    * Mean (mean()): Calculates the mean (average) of values in a group.
    * Median (median()): Calculates the median of values in a group.
    * Minimum (min()): Calculates the minimum value in a group.
    * Maximum (max()): Calculates the maximum value in a group.
    * Count (count()): Counts the number of non-null values in a group.
    * Size (size()): Counts the total number of elements in each group, including null values.
    * Standard Deviation (std()): Calculates the standard deviation of values in a group.
    
9. Sorting:
    * sort_values() method
    * sort_index() method
    
10. concat() function:
    * pd.concat()
    
11. Joins:
    * pd.merge() or merge method too works
    
12. Difference between concat() and merge()

# <code style="background:yellow;color:black">Revision of last lecture:</code>

In [1]:
import numpy as np
import pandas as pd

**Importing a dataset called mckinsey.**

In [2]:
df = pd.read_csv("mckinsey.csv")

**Checking the first 5 rows of the dataset.**

In [3]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

**Checking the datatypes and other info of the dataset.**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   country     1704 non-null   object 
 1   year        1704 non-null   int64  
 2   population  1704 non-null   int64  
 3   continent   1704 non-null   object 
 4   life_exp    1704 non-null   float64
 5   gdp_cap     1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


**Checking the first 10 rows of the dataset.**

In [5]:
df.head(10)

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
6  Afghanistan  1982    12881816      Asia    39.854  978.011439
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
8  Afghanistan  1992    16317921      Asia    41.674  649.341395
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Data slicing:</code>

**In pandas, <code style="background:yellow;color:black">data slicing</code> typically refers to the process of selecting a subset of data from a DataFrame based on certain conditions or criteria. There are several ways to perform data slicing in pandas.**

### <code style="background:yellow;color:black">1. Data slicing through iloc:</code>

* **Format of slicing (implicit index)- df.iloc[rows, columns]**
* **In the following code, we are doing data slicing through iloc on our McKinsey DataFrame df.**
* So if we do df.iloc[1:5, 1:4], the first part before comma refers to rows slice and the second part after comma refers to columns slice, **means we are slicing both rows sand columns at the same time.** 
* iloc 1:5 means implicit index 1 to 4 continuous rows and NOT 5th row. So, rows 1 to 4 are returned and **5th row is excluded.** 
* iloc 1:4 means implicit index "year", "population", "continent" continuous columns and NOT "life_exp" column. So, columns "year", "population" and "continent" are returned and **"life_exp" is excluded.**  

In [6]:
df.iloc[1:5, 1:4]

year  population continent
1  1957     9240934      Asia
2  1962    10267083      Asia
3  1967    11537966      Asia
4  1972    13079460      Asia

### <code style="background:yellow;color:black">2. Data slicing through loc:</code>

* **Format of slicing (explicit index)- df.iloc[rows, columns]**
* **In the following code, we are doing data slicing through loc on our McKinsey DataFrame df.**
* **Slicing with explicit indexing works by <code style="background:yellow;color:black">specifying exact explicit index names as seen on screen.</code>**
* So if we do df.loc[1:5, "year":"continent"], the first part before comma refers to rows slice and the second part after comma refers to columns slice, **means we are slicing both rows sand columns at the same time.**
**Major difference between implicit and explicit slicing is that <code style="background:yellow;color:black">in implicit slicing last index of the slice IS NOT INCLUDED, but in explicit slicing last index of the slice IS INCLUDED.</code>**

**<code style="background:yellow;color:black">Also note that</code> slicing using explicit index will not work with code like df.loc[1:5, 1:4] and will throw error, because rows slicing will work, but in columns, there is no column named 1 or 4. In explicit indexes we have columns called country, year, population, life_exp, etc.** 

In [7]:
df.loc[1:5, "year" : "continent"]

year  population continent
1  1957     9240934      Asia
2  1962    10267083      Asia
3  1967    11537966      Asia
4  1972    13079460      Asia
5  1977    14880372      Asia

In [8]:
# df.loc[1:5, 1:4] 

# Will throw error as we have to specify exact explicit index names of columns as seen on screen

### <code style="background:yellow;color:black">3. Data slicing by extracting one column:</code>

* **In the following code, we are doing data slicing by extracting one particular column from our McKinsey DataFrame df.**

In [9]:
df["country"]

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

### <code style="background:yellow;color:black">4. Data slicing by extracting multiple columns:</code>

* **In the following code, we are doing data slicing by extracting multiple columns from our McKinsey DataFrame df.**
* **<code style="background:yellow;color:black">This is called Multi-indexing.</code>**

In [10]:
df[["country", "continent"]]

country continent
0     Afghanistan      Asia
1     Afghanistan      Asia
2     Afghanistan      Asia
3     Afghanistan      Asia
4     Afghanistan      Asia
...           ...       ...
1699     Zimbabwe    Africa
1700     Zimbabwe    Africa
1701     Zimbabwe    Africa
1702     Zimbabwe    Africa
1703     Zimbabwe    Africa

[1704 rows x 2 columns]

### <code style="background:yellow;color:black">5. Data slicing by combining slicing and multi-indexing:</code>

* **In the following code, we are doing data slicing on our McKinsey DataFrame df, by combining slicing and multi-indexing concepts together.**
* **Advantage of this kind of combining is that we can have any particular column(s) as we want and not just a continuous series of columns as we get in slicing alone.**
* **Both loc and iloc can work and fetch the data slice as required.**
* Below "loc" code is read like this -- Give me all the explicit rows from 1 to 5, and among them give me only 2 columns called year and continent.
* Below "iloc" code is read like this -- Give me all the implicit rows from 1 to 5, and among them give me only 2 columns with implicit indices 1 and 4.

In [11]:
df.loc[1:5, ["year", "continent"]]

year continent
1  1957      Asia
2  1962      Asia
3  1967      Asia
4  1972      Asia
5  1977      Asia

In [12]:
df.iloc[1:5, [1, 4]]

year  life_exp
1  1957    30.332
2  1962    31.997
3  1967    34.020
4  1972    36.088

* **By implicit index slicing i.e. iloc, we can also mention particular rows and columns to be picked or extracted and not continuous series of rows and columns which is done using colon.**
* **Below "iloc" code is read like this -- Give me all the implicit rows with indices 0, 10 and 20 and among them give me only 3 columns with implicit indices 0, 2 and 3.**

In [13]:
df.iloc[[0, 10, 20], [0, 2, 3]]

country  population continent
0   Afghanistan     8425333      Asia
10  Afghanistan    25268405      Asia
20      Albania     3326498    Europe

### <code style="background:yellow;color:black">In slicing, we can also provide step size or jump size, both in iloc and loc:</code>

* **In the following code, we are doing data slicing on our McKinsey DataFrame df, by providing step size both in iloc and loc indexers.**
* **In iloc, as discussed above, last index is EXCLUDED, and in loc last index is INCLUDED.**

In [14]:
df.iloc[1:10:2]

country  year  population continent  life_exp     gdp_cap
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

In [15]:
df.iloc[1:10:2, 1:5:2]

year continent
1  1957      Asia
3  1967      Asia
5  1977      Asia
7  1987      Asia
9  1997      Asia

In [16]:
df.loc[1:11:2]

country  year  population continent  life_exp     gdp_cap
1   Afghanistan  1957     9240934      Asia    30.332  820.853030
3   Afghanistan  1967    11537966      Asia    34.020  836.197138
5   Afghanistan  1977    14880372      Asia    38.438  786.113360
7   Afghanistan  1987    13867957      Asia    40.822  852.395945
9   Afghanistan  1997    22227415      Asia    41.763  635.341351
11  Afghanistan  2007    31889923      Asia    43.828  974.580338

In [17]:
df.loc[1:11:2, "country":"gdp_cap":2]

country  population  life_exp
1   Afghanistan     9240934    30.332
3   Afghanistan    11537966    34.020
5   Afghanistan    14880372    38.438
7   Afghanistan    13867957    40.822
9   Afghanistan    22227415    41.763
11  Afghanistan    31889923    43.828

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Addition of new rows:</code>

**We will learn how to create and add new rows to our existing dataset df. There are two ways of doing this. One way is using the <code style="background:yellow;color:black">append() method</code> and another way is using the <code style="background:yellow;color:black">loc indexer.</code>** 

### <code style="background:yellow;color:black">1. append() method:</code>

**The <code style="background:yellow;color:black">append() method</code> is used to append rows of one DataFrame to another or to add rows  to an existing DataFrame at the end of it. This method is particularly useful when we want to combine two DataFrames along the rows, effectively adding new rows to an existing DataFrame. The append() method returns a new DataFrame with the appended rows and does not modify the original DataFrames.**

**<u>The append() method is used to append rows of one DataFrame to another like shown in the code below:**</u> 

* The ignore_index=True parameter is used to reindex the resulting DataFrame. If not set to True, the index from the appended DataFrame(s) will be retained and it will throw an error. 
* For append(), inplace = True doesnt work because it does not modify the original DataFrames. So, we have to use other way to make the change permanent in case of append() rows. 
* What we can do is create a copy of this code -- df.append(new_row, ignore_index=True), and instead of storing it somewhere else, store it in df itself. We have done exactly this in the code below. So, this way, even though we are not doing inplace = True, in a way we are making that change permanent.

**<code style="background:yellow;color:black">Also note that</code> in the code below, we have converted row to be added to a DataFrame before adding that to our existing DataFrame "df1".**

In [18]:
# Existing DataFrame
data1 = {
    'column1' : [1, 2, 3],
    'column2' : ['A', 'B', 'C']
}
df1 = pd.DataFrame(data1)

# New data to be added
new_data1 = {
    'column1' : [4],
    'column2' : ['D']
}

# Convert new data to a DataFrame
new_df1 = pd.DataFrame(new_data1)

# Append the new data to the existing DataFrame
df1 = df1.append(new_df1, ignore_index = True)

# Display the updated DataFrame
df1.tail()

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_11552\1589937565.py:18: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df1 = df1.append(new_df1, ignore_index = True)


column1 column2
0        1       A
1        2       B
2        3       C
3        4       D

**However in a new code example below, we have not converted row to be added to a DataFrame before adding that to our existing DataFrame. We have directly added the new row to our existing McKinsey DataFrame "df".**

In [19]:
# New data to be added
new_row = {
    'country' : 'India', 
    'year' : 2000, 
    'population' : 13500000, 
    'continent' : 'Asia', 
    'life_exp' : 37.08,  
    'gdp_cap' : 900.23
}

# Append the new data to the existing DataFrame
df = df.append(new_row, ignore_index = True)

# Display the updated DataFrame
df.tail()

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_11552\3212067473.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(new_row, ignore_index = True)


country  year  population continent  life_exp     gdp_cap
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704     India  2000    13500000      Asia    37.080  900.230000

**This code below checks the total number of indices or rows we have in our dataset.** Since index starts from 0, we have total 1705 indices or rows now as we have added one new row to our McKinsey df. 

In [20]:
len(df.index)

1705

### <code style="background:yellow;color:black">2. loc Indexer:</code>

**We can also use the <code style="background:yellow;color:black">loc indexer</code> to add rows to a DataFrame. This approach is more explicit, and we need to specify the index where we want to add the new data.**

**<u>The loc indexer is used to append rows to a DataFrame like shown in the code below:**</u> 
* In this example, iloc[0] is used to access the first (and only) row in the new DataFrame. 
* In loc indexer method, inplace = True doesnt exist, because by default the loc indexer makes the change permanent. 

**<code style="background:yellow;color:black">Also note that</code> in the code below, we have converted row to be added to a DataFrame before adding that to our existing DataFrame "df2". For that reason we are using loc and iloc both indexers.**

In [21]:
# Existing DataFrame
data2 = {
    'column1' : [1, 2, 3],
    'column2' : ['A', 'B', 'C']
}
df2 = pd.DataFrame(data2)

# New data to be added
new_data2 = {
    'column1' : [4],
    'column2' : ['D']
}

# Convert new data to a DataFrame
new_df2 = pd.DataFrame(new_data2)

# Find the index for the new row
new_index = len(df2)

# Use loc to add the new row
df2.loc[new_index] = new_df2.iloc[0]

# Display the updated DataFrame
df2.head()

column1 column2
0        1       A
1        2       B
2        3       C
3        4       D

**However in a new code examples below, we have not converted rows to be added to a DataFrame before adding that to our existing DataFrame. We have directly added the new row to our existing McKinsey DataFrame "df" using just the loc indexer and we haven't used iloc.**

**So in syntax, we are saying at index number 1705, add a new row whose values are supplied as a list. This is very similar to a dictionary, where df.loc[len(df.index)] is <u>dict key</u> and ['India', 2000, 13500000, "Asia", 37.08, 900.23] is <u>dict value.</u>**

In [22]:
df.loc[len(df.index)] = ['India', 2000, 13500000, "Asia", 37.08, 900.23] 

df.tail()

country  year  population continent  life_exp     gdp_cap
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704     India  2000    13500000      Asia    37.080  900.230000
1705     India  2000    13500000      Asia    37.080  900.230000

**In the code below, we can see that one new row has been added to our McKinsey df as number of rows have been increased from 1705 to 1706.**

In [23]:
len(df.index)

1706

**Now in updated df, we have 1705 rows. So, len(df.index) will give 1706 as output, and then we subtract 1 from it to get 1705.**

In [24]:
len(df.index) - 1

1705

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Modification of existing rows:</code>

**We will learn how to modify existing rows to our existing McKinsey DataFrame df. There are two ways of doing this. One way is using the <code style="background:yellow;color:black">loc indexer</code> and another way is using the <code style="background:yellow;color:black">iloc indexer.</code>**

### <code style="background:yellow;color:black">1. Modifying an existing row using loc:</code>

* **Note that before this in our df, in last row we had "year" value as "2000" and now after the code below, in last row "year" value has been changed to "2004".**

In [25]:
df.loc[len(df.index) - 1] = ['India', 2004, 13500000, "Asia", 37.08, 900.23] 

df.tail()

country  year  population continent  life_exp     gdp_cap
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704     India  2000    13500000      Asia    37.080  900.230000
1705     India  2004    13500000      Asia    37.080  900.230000

### <code style="background:yellow;color:black">2. Modifying or updating an existing row using iloc:</code> 

* **Note that before this in our df, in last row we had "year" value as "2004" and now after the code below, in last row "year" value has been changed to "2008". In this case, we are modifying last row. If we want to modify second last row we can say df.iloc[len(df.index) - 2].**

In [26]:
df.iloc[len(df.index) - 1] = ['India', 2008, 13500000, "Asia", 37.08, 900.23] 

df.tail()

country  year  population continent  life_exp     gdp_cap
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704     India  2000    13500000      Asia    37.080  900.230000
1705     India  2008    13500000      Asia    37.080  900.230000

* **<code style="background:yellow;color:black">In the code below, we are adding 4 rows at the bottom to the existing rows permanently</code>**
* We are adding one row at a time at the bottom. 
* First, first line of code will get executed, so that particular new row will be added at index 1706, then second line of code will get executed and that particular new row will be added at index 1707 and so on.

In [27]:
df.loc[len(df.index)] = ['India',2000,13500000,"asia",37.08,900.23]
df.loc[len(df.index)] = ['Sri Lanka',2022,130000000,"asia",80.00,500.00]
df.loc[len(df.index)] = ['Sri Lanka',2022,130000000,"asia",80.00,500.00]
df.loc[len(df.index)] = ['India',2000,13500000,"asia",80.00,900.23]
df.tail(10)

country  year  population continent  life_exp     gdp_cap
1700   Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701   Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702   Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703   Zimbabwe  2007    12311143    Africa    43.487  469.709298
1704      India  2000    13500000      Asia    37.080  900.230000
1705      India  2008    13500000      Asia    37.080  900.230000
1706      India  2000    13500000      asia    37.080  900.230000
1707  Sri Lanka  2022   130000000      asia    80.000  500.000000
1708  Sri Lanka  2022   130000000      asia    80.000  500.000000
1709      India  2000    13500000      asia    80.000  900.230000

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Duplicate rows:</code>

**<code style="background:yellow;color:black">Duplicate rows</code> in a pandas DataFrame are rows that have identical values in all columns, means everything every column value same except for index. These rows can arise from errors in data collection, merging datasets, or other data manipulation operations. It's important to identify and handle duplicate rows to ensure the accuracy of your analysis.**

### <code style="background:yellow;color:black">Checking for Duplicate Rows:</code>

**We can use the <code style="background:yellow;color:black">duplicated() method</code> in pandas to identify duplicate rows in a DataFrame. This method returns a boolean Series indicating whether each row is a duplicate of a previous row. "True" means that row is duplicate.**

**<u>The duplicated() method is used to identify duplicate rows in our McKinsey DataFrame like shown in the code below:**</u> 
* Now among all the new rows added to the original df in above cell, are there any duplicate rows? -- YES.
* Index of duplicate row is 1708, its a duplicate of row 1707. 
* Please note in rows 1704 and 1706, there is one minor difference of case sensitiveness of "Asia" and "asia".

**<code style="background:yellow;color:black">Also note that</code> in the code below, we did not get true for 1707, because that row is the first occurence in the dataset, though 1707 and 1708 are same. So, first occurence is not duplicate, second, third, fourth and so on are duplicates.**

In [28]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1705    False
1706    False
1707    False
1708     True
1709    False
Length: 1710, dtype: bool

**Below is another example of identification of duplicate rows using duplicated() method.**

In [29]:
# Sample DataFrame
data3 = {
    'column1' : [1, 2, 3, 4, 2],
    'column2' : ['A', 'B', 'C', 'D', 'B']
}

df3 = pd.DataFrame(data3)

# Check for duplicate rows
df3.duplicated()

0    False
1    False
2    False
3    False
4     True
dtype: bool

**<code style="background:yellow;color:black">If we want to store or filter out all the duplicate rows, we will use the concept of masking.</code> We will pass df.duplicated() into df.loc. We will get all the rows where it is "True".** We are using McKinsey dataset.

In [30]:
# Check and store duplicate rows
df_dup = df.loc[df.duplicated()]

# Display duplicate rows
df_dup

country  year  population continent  life_exp  gdp_cap
1708  Sri Lanka  2022   130000000      asia      80.0    500.0

**In the code example of sample dataset below, we are storing directly duplicate rows in a variable, and <code style="background:yellow;color:black">not using df.loc.</code>** 

In [31]:
# Check and store duplicate rows
duplicate_rows = df3.duplicated()

# Display duplicate rows
df3[duplicate_rows]

column1 column2
4        2       B

### <code style="background:yellow;color:black">Deleting Duplicate Rows:</code>

**To remove or delete duplicate rows from a DataFrame, we can use the <code style="background:yellow;color:black">drop_duplicates() method.</code> This method returns a DataFrame with duplicate rows removed. This method does not modify the original DataFrame in place. Instead, it returns a new DataFrame with duplicate rows removed. If we want to make the change permanent, we need to either assign the result back to the original DataFrame or use the inplace = True parameter.** 

**<u>The drop_duplicates() method is used to delete duplicate rows in our McKinsey DataFrame like shown in the code below:**</u>
* We can see 1708 is not there in the output.
* Will this change be permanent? -- NO, because we have not written inplace = True. We are checking using df.tail() in next cell.

In [32]:
df.drop_duplicates()

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2008    13500000      Asia    37.080  900.230000
1706        India  2000    13500000      asia    37.080  900.230000
1707    Sri Lanka  2022   130000000      asia    80.000  500.000000
1709        India  2000    13500000      asia    80.000  900.230000

[1709 rows x 6 columns]

In [33]:
df.tail() # Change is not permanent

country  year  population continent  life_exp  gdp_cap
1705      India  2008    13500000      Asia     37.08   900.23
1706      India  2000    13500000      asia     37.08   900.23
1707  Sri Lanka  2022   130000000      asia     80.00   500.00
1708  Sri Lanka  2022   130000000      asia     80.00   500.00
1709      India  2000    13500000      asia     80.00   900.23

* **<code style="background:yellow;color:black">By default, drop_duplicates() keeps the first occurrence of each set of duplicate rows and removes subsequent occurrences.</code> If we want to keep the last occurrence instead, we can use the keep = "last" parameter.**
* In above df.tail() cell, we can see 1707 and 1708 are duplicates.
* So, in the code below, we will keep the first occurence while deleting all the other duplicate occurences. By default if we don't pass keep parameter, it will automatically delete all the remaining occurences as seen in the above code cell. But for code readability we pass keep = "first" parameter, means keep the first occurence and remove all other occurences of it.
* Will this change be permanent? -- NO, because we have not written inplace = True. We are checking using df.tail() in next cell.

In [34]:
df.drop_duplicates(keep = "first")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2008    13500000      Asia    37.080  900.230000
1706        India  2000    13500000      asia    37.080  900.230000
1707    Sri Lanka  2022   130000000      asia    80.000  500.000000
1709        India  2000    13500000      asia    80.000  900.230000

[1709 rows x 6 columns]

In [35]:
df.tail() # Change is not permanent

country  year  population continent  life_exp  gdp_cap
1705      India  2008    13500000      Asia     37.08   900.23
1706      India  2000    13500000      asia     37.08   900.23
1707  Sri Lanka  2022   130000000      asia     80.00   500.00
1708  Sri Lanka  2022   130000000      asia     80.00   500.00
1709      India  2000    13500000      asia     80.00   900.23

* In above df.tail(), we can see 1707 and 1708 are duplicates. 
* **Now, in the code below, we will <code style="background:yellow;color:black">keep the last occurence while deleting all the other previous duplicate occurences</code>.** So, automatically it will delete all the remaining occurences.
* As we can see in code below, it has removed 1707, because 1707 was the first occurence. keep = "last' means keep the last occurence and remove all other occurences before of it.
* Will this change be permanent? -- NO, because we have not written inplace = True. We are checking using df.tail() in next cell.

In [36]:
df.drop_duplicates(keep = "last")

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1704        India  2000    13500000      Asia    37.080  900.230000
1705        India  2008    13500000      Asia    37.080  900.230000
1706        India  2000    13500000      asia    37.080  900.230000
1708    Sri Lanka  2022   130000000      asia    80.000  500.000000
1709        India  2000    13500000      asia    80.000  900.230000

[1709 rows x 6 columns]

In [37]:
df.tail() # Change is not permanent

country  year  population continent  life_exp  gdp_cap
1705      India  2008    13500000      Asia     37.08   900.23
1706      India  2000    13500000      asia     37.08   900.23
1707  Sri Lanka  2022   130000000      asia     80.00   500.00
1708  Sri Lanka  2022   130000000      asia     80.00   500.00
1709      India  2000    13500000      asia     80.00   900.23

**Now in the code cell below, we will keep the first occuring duplicate row among all duplicate rows while deleting all other occuerences and <code style="background:yellow;color:black">make that change permanent by passing inplace = True parameter.</code> We are checking the result using df.tail() in next cell to see permanent change i.e first duplicate row of index 1707 is retained while row with index 1708 is permanently deleted.**

In [38]:
df.drop_duplicates(keep = "first", inplace = True)

df.tail() # Change is now permanent that means our dataframe is permanently modified

country  year  population continent  life_exp  gdp_cap
1704      India  2000    13500000      Asia     37.08   900.23
1705      India  2008    13500000      Asia     37.08   900.23
1706      India  2000    13500000      asia     37.08   900.23
1707  Sri Lanka  2022   130000000      asia     80.00   500.00
1709      India  2000    13500000      asia     80.00   900.23

**Since we have made permanent changes of addition and deletion of few rows in our McKinsey DataFrame, we are again loading the DataFrame by the code below so that we can further learn new concepts on this same DataFrame.**

In [39]:
df = pd.read_csv("mckinsey.csv")

df.tail(10)

country  year  population continent  life_exp     gdp_cap
1694  Zimbabwe  1962     4277736    Africa    52.358  527.272182
1695  Zimbabwe  1967     4995432    Africa    53.995  569.795071
1696  Zimbabwe  1972     5861135    Africa    55.635  799.362176
1697  Zimbabwe  1977     6642107    Africa    57.674  685.587682
1698  Zimbabwe  1982     7636524    Africa    60.363  788.855041
1699  Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Deletion of rows and columns:</code>

### <code style="background:yellow;color:black">drop() method:</code>

**In Pandas, we can delete rows and columns from a DataFrame <code style="background:yellow;color:black">using the drop method.</code> The drop method allows us to remove specific rows or columns by specifying their labels or indices. Here's how we can delete rows and columns in Pandas:**

### <code style="background:yellow;color:black">1. Deleting a single row:</code> 

* Here in our McKinsey df, we are deleting 3rd index row. 
* As we know that vertical axis is 1 and horizontal axis is 0, so we have provided axis = 0 as a parameter. If we don't provide axis = 0, Pandas will assume the default value (axis = 0), but it's good practice to be explicit about our intentions to improve code readability and avoid potential confusion. 
* Also note that this change is not permanent, means in original dataset row is not deleted because we havent provided the parameter inplace = True. This we have checked with df.head() in next cell.

In [40]:
df.drop(3, axis = 0)

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
5     Afghanistan  1977    14880372      Asia    38.438  786.113360
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1703 rows x 6 columns]

In [41]:
df.head() # Change is not permanent

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

### <code style="background:yellow;color:black">2. Deleting multiple rows:</code>

* To delete multiple rows at a time, instead of a number, we have to provide a list of indexes of those rows which we want to delete.
* **Important: Anywhere, if we want to delete or access one thing, we have to write a single number or index as a parameter in the method, and, if we want to delete or access multiple things, we have to write a list as a parameter.**
* Here in our McKinsey df, we are deleting 2nd, 4th and 5th index rows.
* Also note that this change is not permanent, means in original dataset row is not deleted because we havent provided the parameter inplace = True. This we have checked with df.head() in next cell.

In [42]:
df.drop([2,4,5], axis = 0)

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
6     Afghanistan  1982    12881816      Asia    39.854  978.011439
7     Afghanistan  1987    13867957      Asia    40.822  852.395945
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1701 rows x 6 columns]

In [43]:
df.head(10) # Change is not permanent

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
6  Afghanistan  1982    12881816      Asia    39.854  978.011439
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
8  Afghanistan  1992    16317921      Asia    41.674  649.341395
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

**In this another example code below, rows with indices 1 and 3 will be removed from the DataFrame.**

In [44]:
# Sample DataFrame
data4 = {
    'A' : [1, 2, 3, 4],
    'B' : [5, 6, 7, 8],
    'C' : [9, 10, 11, 12]
}

df4 = pd.DataFrame(data4)

# Delete rows with index 1 and 3
df4.drop([1, 3])

A  B   C
0  1  5   9
2  3  7  11

In [45]:
df4.head() # Change is not permanent

A  B   C
0  1  5   9
1  2  6  10
2  3  7  11
3  4  8  12

### <code style="background:yellow;color:black">3. Deleting a single column:</code>

* Here in our McKinsey df, we are deleting "country" column. 
* As we know that vertical axis is 1 and horizontal axis is 0, so we have provided axis = 1 as a parameter. If we don't provide axis = 1, Pandas will assume the default value (axis = 0), but it's good practice to be explicit about our intentions to improve code readability and avoid potential confusion. 
* Also note that this change is not permanent, means in original dataset row is not deleted because we havent provided the parameter inplace = True. This we have checked with df.head() in next cell.

In [46]:
df.drop("country", axis = 1) 

year  population continent  life_exp     gdp_cap
0     1952     8425333      Asia    28.801  779.445314
1     1957     9240934      Asia    30.332  820.853030
2     1962    10267083      Asia    31.997  853.100710
3     1967    11537966      Asia    34.020  836.197138
4     1972    13079460      Asia    36.088  739.981106
...    ...         ...       ...       ...         ...
1699  1987     9216418    Africa    62.351  706.157306
1700  1992    10704340    Africa    60.377  693.420786
1701  1997    11404948    Africa    46.809  792.449960
1702  2002    11926563    Africa    39.989  672.038623
1703  2007    12311143    Africa    43.487  469.709298

[1704 rows x 5 columns]

In [47]:
df.head() # Change is not permanent

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

### <code style="background:yellow;color:black">4. Deleting multiple columns:</code>

* To delete multiple columns at a time, instead of a number, we have to provide a list of column names which we want to delete.
* **Important: Anywhere, if we want to delete or access one thing, we have to write a single number or index as a parameter in the method, and, if we want to delete or access multiple things, we have to write a list as a parameter.**
* Here in our McKinsey df, we are deleting "country", "year" and "population" columns.
* Also note that this change is not permanent, means in original dataset row is not deleted because we havent provided the parameter inplace = True. This we have checked with df.head() in next cell.

In [48]:
df.drop(["country", "year", "population"], axis = 1) 

continent  life_exp     gdp_cap
0         Asia    28.801  779.445314
1         Asia    30.332  820.853030
2         Asia    31.997  853.100710
3         Asia    34.020  836.197138
4         Asia    36.088  739.981106
...        ...       ...         ...
1699    Africa    62.351  706.157306
1700    Africa    60.377  693.420786
1701    Africa    46.809  792.449960
1702    Africa    39.989  672.038623
1703    Africa    43.487  469.709298

[1704 rows x 3 columns]

In [49]:
df.head() # Change is not permanent

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

**In this another example code below, columns "B" and "C" will be removed from the DataFrame.**

In [50]:
# Sample DataFrame
data5 = {
    'A' : [1, 2, 3, 4],
    'B' : [5, 6, 7, 8],
    'C' : [9, 10, 11, 12]
}

df5 = pd.DataFrame(data5)

# Delete columns 'B' and 'C'
df5.drop(['B', 'C'], axis = 1)

A
0  1
1  2
2  3
3  4

In [51]:
df5.head() # Change is not permanent

A  B   C
0  1  5   9
1  2  6  10
2  3  7  11
3  4  8  12

### <code style="background:yellow;color:black">5. Deleting rows and columns permanently:</code>

* By default, the drop method returns a new DataFrame with the specified rows or columns removed, leaving the original DataFrame unchanged. 
* But if we want to modify the DataFrame in-place, we can use the inplace=True parameter
* In code examples below, we are deleting rows and columns and checking the change is permanent with df.head() in next cell.
* Keep in mind that using inplace = True can modify the original DataFrame directly, so use it with caution.

In [52]:
# Delete rows with index 1 and 3 in-place
df4.drop([1, 3], axis = 0, inplace = True)

# Delete columns 'B' and 'C' in-place
df5.drop(['B', 'C'], axis = 1, inplace = True)

In [53]:
df4.head() # df4 is modified in place permanently

A  B   C
0  1  5   9
2  3  7  11

In [54]:
df5.head() # df5 is modified in place permanently

A
0  1
1  2
2  3
3  4

In [55]:
# Delete rows with index 1, 3 and 5 in-place of McKinsey dataset
df.drop([1, 3, 5], axis = 0, inplace = True)

# Delete columns "country" and "year" in-place of McKinsey dataset
df.drop(["country", "year"], axis = 1, inplace = True)

In [56]:
df.head(10)

population continent  life_exp      gdp_cap
0      8425333      Asia    28.801   779.445314
2     10267083      Asia    31.997   853.100710
4     13079460      Asia    36.088   739.981106
6     12881816      Asia    39.854   978.011439
7     13867957      Asia    40.822   852.395945
8     16317921      Asia    41.674   649.341395
9     22227415      Asia    41.763   635.341351
10    25268405      Asia    42.129   726.734055
11    31889923      Asia    43.828   974.580338
12     1282697    Europe    55.230  1601.056136

**Since we have made permanent changes of deletion of few rows and columns in our McKinsey DataFrame, we are again loading the DataFrame by the code below so that we can further learn new concepts on this same DataFrame.**

In [57]:
df = pd.read_csv("mckinsey.csv")

df.head(10)

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
6  Afghanistan  1982    12881816      Asia    39.854  978.011439
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
8  Afghanistan  1992    16317921      Asia    41.674  649.341395
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Resetting index:</code>

### <code style="background:yellow;color:black">reset_index() method:</code>

**The <code style="background:yellow;color:black">reset_index() method</code> in pandas is used to reset the index of a DataFrame. When we perform operations that result in changes to the DataFrame's structure (such as filtering, grouping, or merging), the index may become irregular or modified. The reset_index() method allows us to revert the index changes and create a new default integer-based index. reset_index() makes the indexes reset back to original state. And it also creates a column called "index".**

* **Following code illustrates reset_index() method.**
* **By default, reset_index() adds a new integer-based index and keeps the existing index too.**
* **If we want to remove the existing index as a column in the DataFrame, we can use the <code style="background:yellow;color:black">drop = True parameter.</code>**
* **If we want to make the change permanent in the original DataFrame df at the same time, we can use the <code style="background:yellow;color:black">inplace = True parameter.</code>**
* We are checking these changes using df.head().

In [58]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [59]:
df.reset_index()

index      country  year  population continent  life_exp     gdp_cap
0         0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1         1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2         2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3         3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4         4  Afghanistan  1972    13079460      Asia    36.088  739.981106
...     ...          ...   ...         ...       ...       ...         ...
1699   1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700   1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701   1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702   1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703   1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 7 columns]

In [60]:
df.reset_index(drop = True, inplace = True)

df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Aggregate functions:</code>

* **We have learnt aggregate functions in numpy, same can be applied in pandas as well.**
* **In pandas, <code style="background:yellow;color:black">aggregate functions (often referred to as aggregation functions)</code> are functions that perform a computation on a set of values. These functions are typically applied to a group of data points, often grouped by one or more categorical variables.** 
* **When we apply an aggregate function directly to a column, it will compute the aggregate value for the entire column.**
* **The result is a summary statistic or a new set of values that represent the characteristics of the grouped data.**

**<code style="background:yellow;color:black">Some common aggregate functions in pandas include:**</code>

* **Sum (sum()):** Calculates the sum of values in a group.
* **Mean (mean()):** Calculates the mean (average) of values in a group.
* **Median (median()):** Calculates the median of values in a group.
* **Minimum (min()):** Calculates the minimum value in a group.
* **Maximum (max()):** Calculates the maximum value in a group.
* **Count (count()):** Counts the number of non-null values in a group.
* **Size (size()):** Counts the total number of elements in each group, including null values.
* **Standard Deviation (std()):** Calculates the standard deviation of values in a group.

**<code style="background:yellow;color:black">Also note that</code> if we want to apply multiple aggregate functions at once, we can use <code style="background:yellow;color:black">the agg() method,</code> which will be discussed in next sesion.**

**<code style="background:yellow;color:black">In this example below,</code> aggregate functions like mean(), sum(), min(), max(), median(), and std() are applied directly to the 'column1' without using groupby. These functions compute the respective summary statistics for the entire column.**

In [61]:
# Sample DataFrame
data6 = {
    'column1': [1, 2, 3, 4, 2, 5, 3, 1]
}
df6 = pd.DataFrame(data6)

# Apply aggregate functions directly to the column
mean_value = df6['column1'].mean()
sum_value = df6['column1'].sum()
min_value = df6['column1'].min()
max_value = df6['column1'].max()
median_value = df6['column1'].median()
std_deviation_value = df6['column1'].std()

# Display the results
print(f"Mean: {mean_value}")
print(f"Sum: {sum_value}")
print(f"Min: {min_value}")
print(f"Max: {max_value}")
print(f"Median: {median_value}")
print(f"Standard Deviation: {std_deviation_value}")

Mean: 2.625
Sum: 21
Min: 1
Max: 5
Median: 2.5
Standard Deviation: 1.407885953173359


* **<code style="background:yellow;color:black">In this example below,</code> we are applying aggregate functions like mean(), sum(), min(), max(), median(), and std() to the "life_expectancy" column of our McKinsey DataFrame.** 
* **These functions compute the respective summary statistics for the entire column.** 
* We can have two different syntaxes to do this task as shown in below code cells. 
* In one syntax, we store aggregate values in a variable and in another syntax, we apply the aggregate functions directly on the column as required. 

In [62]:
# Apply aggregate functions directly to the column
mean_value = df["life_exp"].mean()
sum_value = df["life_exp"].sum()
min_value = df["life_exp"].min()
max_value = df["life_exp"].max()
median_value = df["life_exp"].median()
std_deviation_value = df["life_exp"].std()

# Display the results
print(f"Mean: {mean_value}")
print(f"Sum: {sum_value}")
print(f"Min: {min_value}")
print(f"Max: {max_value}")
print(f"Median: {median_value}")
print(f"Standard Deviation: {std_deviation_value}")

Mean: 59.474439366197174
Sum: 101344.44467999999
Min: 23.599
Max: 82.603
Median: 60.7125
Standard Deviation: 12.917107415241187


In [63]:
life_expectancy = df["life_exp"]

# Apply aggregate functions directly to the column
mean_value = life_expectancy.mean()
sum_value = life_expectancy.sum()
min_value = life_expectancy.min()
max_value = life_expectancy.max()
median_value = life_expectancy.median()
std_deviation_value = life_expectancy.std()

# Display the results
print(f"Mean: {mean_value}")
print(f"Sum: {sum_value}")
print(f"Min: {min_value}")
print(f"Max: {max_value}")
print(f"Median: {median_value}")
print(f"Standard Deviation: {std_deviation_value}")

Mean: 59.474439366197174
Sum: 101344.44467999999
Min: 23.599
Max: 82.603
Median: 60.7125
Standard Deviation: 12.917107415241187


**Internally in pandas, mean is calculated like the code implementation below:**

In [64]:
count_le = life_expectancy.count()
sum_le = life_expectancy.sum()

sum_le / count_le

59.474439366197174

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Sorting:</code>

**Sorting in pandas refers to arranging the rows or columns of a DataFrame in a specific order based on the values in one or more columns. Sorting can be done in ascending or descending order. The primary method for sorting in pandas is <code style="background:yellow;color:black">sort_values() for sorting by values, and sort_index() for sorting by index.</code>**

### <code style="background:yellow;color:black">1. sort_values() method:</code>

* **To sort a DataFrame by the values in one or more columns, we can use the <code style="background:yellow;color:black">sort_values() method.</code> Means we can also provide a list of column names to be sorted together.** 
* **<code style="background:yellow;color:black">This is very similar to ORDER BY clause in sql.</code>**
* Here in the code below, we are using sort_values() method on "life_exp" column of our McKinsey DataFrame df.
* "life_exp" column is sorted in increasing order. Rows are shuffled and so are the indices. Same as SQL, entire rows have moved up and down.

In [65]:
df.sort_values(["life_exp"])

country  year  population continent  life_exp       gdp_cap
1292            Rwanda  1992     7290203    Africa    23.599    737.068595
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
...                ...   ...         ...       ...       ...           ...
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
802              Japan  2002   127065841      Asia    82.000  28604.591900
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1704 rows x 6 columns]

* **<code style="background:yellow;color:black">If we want to sort columns in descending order, we have to pass a parameter called "ascending" as shown below.</code>** 
* **By default ascending is True, just like sql. Now we have sorted "life_exp" column in descending order.** 

In [66]:
df.sort_values(["life_exp"], ascending = False) 

country  year  population continent  life_exp       gdp_cap
803              Japan  2007   127467972      Asia    82.603  31656.068060
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
802              Japan  2002   127065841      Asia    82.000  28604.591900
695            Iceland  2007      301931    Europe    81.757  36180.789190
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
...                ...   ...         ...       ...       ...           ...
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
36              Angola  1952     4232095    Africa    30.015   3520.610273
552             Gambia  1952      284320    Africa    30.000    485.230659
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
1292            Rwanda  1992     7290203    Africa    23.599    737.068595

[1704 rows x 6 columns]

* **<code style="background:yellow;color:black">We can sort multiple columns at the same time.</code>**
* **Here in code below, first pandas will sort in ascending order of "year" column and whenever there is a tie it will sort on the ascending order of "life_exp" column, same as sql.**
* **Also note that, in year 1952, all life expectancy is sorted in ascending order and so on for every year.** 

In [67]:
df.sort_values(["year", "life_exp"])

country  year  population continent  life_exp       gdp_cap
0          Afghanistan  1952     8425333      Asia    28.801    779.445314
552             Gambia  1952      284320    Africa    30.000    485.230659
36              Angola  1952     4232095    Africa    30.015   3520.610273
1344      Sierra Leone  1952     2143249    Africa    30.331    879.787736
1032        Mozambique  1952     6446316    Africa    31.286    468.526038
...                ...   ...         ...       ...       ...           ...
71           Australia  2007    20434176   Oceania    81.235  34435.367440
1487       Switzerland  2007     7554661    Europe    81.701  37506.419070
695            Iceland  2007      301931    Europe    81.757  36180.789190
671   Hong Kong, China  2007     6980412      Asia    82.208  39724.978670
803              Japan  2007   127467972      Asia    82.603  31656.068060

[1704 rows x 6 columns]

* **In sql we can write - order by [col_name] asc, [col_name] desc.**
* **Same we can do in pandas as well, <code style="background:yellow;color:black">we can sort one column in ascending and one column in desending as shown below.</code>**

In [68]:
df.sort_values(["year", "life_exp"], ascending = [True, False])

country  year  population continent  life_exp       gdp_cap
1140        Norway  1952     3327728    Europe    72.670  10095.421720
684        Iceland  1952      147962    Europe    72.490   7267.688428
1080   Netherlands  1952    10381988    Europe    72.130   8941.571858
1464        Sweden  1952     7124673    Europe    71.860   8527.844662
408        Denmark  1952     4334000    Europe    70.780   9692.385245
...            ...   ...         ...       ...       ...           ...
887        Lesotho  2007     2012649    Africa    42.592   1569.331442
1355  Sierra Leone  2007     6144562    Africa    42.568    862.540756
1691        Zambia  2007    11746035    Africa    42.384   1271.211593
1043    Mozambique  2007    19951656    Africa    42.082    823.685621
1463     Swaziland  2007     1133066    Africa    39.613   4513.480643

[1704 rows x 6 columns]

### <code style="background:yellow;color:black">2. sort_index() method:</code>

* **To sort a DataFrame by its index, we can use the <code style="background:yellow;color:black">sort_index() method.</code> By default, it sorts in ascending order, but we can specify "ascending = False" to sort in descending order.**
* Here in the code below, we are using sort_index() method on sample DataFrame df7.

In [69]:
# Sample DataFrame
data7 = {
    'column1' : [3, 1, 4, 2],
    'column2' : ['D', 'B', 'A', 'C']
}
df7 = pd.DataFrame(data7)

df7

column1 column2
0        3       D
1        1       B
2        4       A
3        2       C

In [70]:
# Sort by index in descending order
df7_sorted_index = df7.sort_index(ascending = False)

# Display the DataFrame with sorted index
print("Sorted by index:")
print(df7_sorted_index)

Sorted by index:
   column1 column2
3        2       C
2        4       A
1        1       B
0        3       D


* **Both sort_values() and sort_index() return a new DataFrame with the sorted order; they do not modify the original DataFrame in place.** 
* **<code style="background:yellow;color:black">If we want to modify the original DataFrame, we have to use the "inplace = True" parameter.</code>**
* As we can see in code cell below, df7 is not changed. Then we sort values in "column1" and make the change permanent with "inplace = True" parameter.

In [71]:
df7

column1 column2
0        3       D
1        1       B
2        4       A
3        2       C

In [11]:
df7.sort_values("column1", inplace = True)

df7

column1 column2
1        1       B
3        2       C
0        3       D
2        4       A

<hr style="border: 1px solid gray;">

## <code style="background:yellow;color:black">concat() function:</code>

* **In pandas, <code style="background:yellow;color:black">the concat() function</code> is used to concatenate two or more pandas objects along a particular axis. This can include DataFrames or Series.** 
* **The concat() function provides more flexibility than the append() method because it allows us to concatenate along either rows (axis 0) or columns (axis 1) and also provides options for handling indices.**
* **<code style="background:yellow;color:black">concat() function performs similar opearion as union (as seen in sql), but it can concatenate two DataFrames horizontally as well.</code>**
* **<code style="background:yellow;color:black">Here's a basic example of using concat() to concatenate two DataFrames <u>along rows.</u></code>**
* In this example, pd.concat([df1, df2], ignore_index = True) concatenates df8 and df9 along rows, and the resulting DataFrame (result_df) has a new, continuous index.

In [72]:
# DataFrame 1
data8 = {
    'column1' : [1, 2, 3],
    'column2' : ['A', 'B', 'C']
}
df8 = pd.DataFrame(data8)

# DataFrame 2
data9 = {
    'column1' : [4, 5],
    'column2' : ['D', 'E']
}
df9 = pd.DataFrame(data9)

# Concatenate along rows (axis = 0)
result_df1 = pd.concat([df8, df9], axis = 0, ignore_index = True)

# Display the result
result_df1

column1 column2
0        1       A
1        2       B
2        3       C
3        4       D
4        5       E

**<code style="background:yellow;color:black">Here's an example of using concat() function to concatenate two DataFrames <u>along columns</u> by specifying "axis = 1".</code>**

In [73]:
df8

column1 column2
0        1       A
1        2       B
2        3       C

In [74]:
df9

column1 column2
0        4       D
1        5       E

In [75]:
# Concatenate along columns (axis = 1)
result_df2 = pd.concat([df8, df9], axis = 1, ignore_index = True)

# Display the result
result_df2

0  1    2    3
0  1  A  4.0    D
1  2  B  5.0    E
2  3  C  NaN  NaN

* **Below is another example illustrating concat() function.**
* **We are creating a new dataframe called users. This df has 3 users with their name.**
* **We are creating another new dataframe called msgs. This df has messages stored by particular user.** Lets say these are whatsapp msgs. 
* **Both these df have userid as a common column between them.**
* In sql we have **joins**, which means joining two tables horizontally over a commom column between two data tables, and we have also learnt **union**, which means joining or staking two tables on top of each other vertically.
* If we do execute below code, means we are passing two dataframes in concat() method, and we want to merge them vertically, so we pass parameter "axis = 1", we get the union of two dataframes.
* **<code style="background:yellow;color:black">Also note NaN.</code> It means not a number or a missing values. In first df, there were no msg vales, and in 2nd df, there were no name values.**
* **<code style="background:yellow;color:black">Also note that</code> the indices are also as it is, i.e. 0,1,2 and 0,1,2,3.**

In [76]:
users = pd.DataFrame(
    {
        "userid" : [1, 2, 3], 
        "name" : ["sharadh", "shahid", "khushali"]
    }
)

users

userid      name
0       1   sharadh
1       2    shahid
2       3  khushali

In [77]:
msgs = pd.DataFrame(
    {
        "userid" : [1, 1, 2, 4], 
        "msg" : ['hmm', "acha", "theek hai", "nice"]
    }
)

msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

In [78]:
pd.concat([users, msgs], axis = 0)

userid      name        msg
0       1   sharadh        NaN
1       2    shahid        NaN
2       3  khushali        NaN
0       1       NaN        hmm
1       1       NaN       acha
2       2       NaN  theek hai
3       4       NaN       nice

**If we want to make indexing proper, means continuos indices, we can pass ignore_index = True parameter.**

In [79]:
pd.concat([users, msgs], axis = 0, ignore_index = True)

userid      name        msg
0       1   sharadh        NaN
1       2    shahid        NaN
2       3  khushali        NaN
3       1       NaN        hmm
4       1       NaN       acha
5       2       NaN  theek hai
6       4       NaN       nice

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Joins:</code>

**In pandas, <code style="background:yellow;color:black">joins refer to the operation of combining two or more DataFrames based on common columns or indices.</code> This operation is similar to the SQL JOIN operation and allows us to merge data from different DataFrames into a single DataFrame, aligning rows based on specified key columns or indices.**

### <code style="background:yellow;color:black">merge() method:</code>

**The primary method for performing joins in pandas is the <code style="background:yellow;color:black">merge() method().</code> The merge() method provides a flexible and powerful way to combine DataFrames. It is similar to SQL joins and provides a powerful way to bring together data from different sources into a single DataFrame.**

**<u>The merge() method is used to merge or join two DataFrames over a common column between them, like shown in the code below:**</u> 
* **Here we pass a parameter called "on", which means common column on which we perform join.**
* **<code style="background:yellow;color:black">If we dont write anything else than this it will be inner join.</code>**
* We have learnt all types of joins in sql. So we will focus on learning syntax of joins.

In [64]:
users

userid      name
0       1   sharadh
1       2    shahid
2       3  khusalli

In [65]:
msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

In [66]:
users.merge(msgs, on = "userid")

userid     name        msg
0       1  sharadh        hmm
1       1  sharadh       acha
2       2   shahid  theek hai

* **<code style="background:yellow;color:black">If we pass the parameter how = "left", it will perform left join.</code>** 
* Here, users is our left table and msgs which is inside parathesis is our right table.
* In sql, we get null for missing vales, in pandas we get NaN. And both are same only.

In [67]:
users.merge(msgs, on = "userid", how = "left")

userid      name        msg
0       1   sharadh        hmm
1       1   sharadh       acha
2       2    shahid  theek hai
3       3  khusalli        NaN

* **<code style="background:yellow;color:black">If we pass the parameter how = "right", it will perform right join.</code>**
* Here, users is our left table and msgs which is inside parathesis is our right table.
* **One important info- If we write "inner" in how =, we can get inner join, and if we write "outer" in how =, we can get outer join.**
* **<code style="background:yellow;color:black">If we want to do self join we write users.merge(users, ....). But we hardly require this self join.</code>**

In [68]:
users.merge(msgs, on = "userid", how = "right")

userid     name        msg
0       1  sharadh        hmm
1       1  sharadh       acha
2       2   shahid  theek hai
3       4      NaN       nice

* **<code style="background:yellow;color:black">What if the column names are different but mean the same thing, for example- "userid" and "id".</code>**
* Below we are changing or renaming "userid" column to "id", and we are also making the change permanent by using inplace = True. And for that we use **rename()** method. 
* We are doing this to learn the syntax of joins when column names are different but mean the same thing.

In [69]:
users.rename(columns = {"userid" : "id"}, inplace = True)

In [70]:
users.head()

id      name
0   1   sharadh
1   2    shahid
2   3  khusalli

In [71]:
msgs

userid        msg
0       1        hmm
1       1       acha
2       2  theek hai
3       4       nice

* **<code style="background:yellow;color:black">We will learn how to do join them now when we dont have common columns.</code>**
* **So, syntax is instead of "on", we write left_on = "id" and right_on = "userid". We are essentially saying from the left table take "id" as common column and from the right table take "userid" as the common column. An this is inner join because we are not providing how = parameter and by default it means inner join.** 
* The code is implemented below:

In [72]:
users.merge(msgs, left_on = "id", right_on = "userid") # inner join

id     name  userid        msg
0   1  sharadh       1        hmm
1   1  sharadh       1       acha
2   2   shahid       2  theek hai

### <code style="background:yellow;color:black">Difference between concat() and merge():</code>

**While both concat and merge methods in pandas can be used for combining DataFrames, they serve different purposes and are used in different scenarios:**

**<code style="background:yellow;color:black">Concatenation (concat):**</code>

* Concatenation is used for combining DataFrames along a particular axis (either rows or columns).
* It is useful when we want to combine DataFrames that have the same columns or the same index along the concatenation axis.
* Concatenation is simple and efficient for stacking DataFrames.

**<code style="background:yellow;color:black">Merging (merge):**</code>

* Merging is used for combining DataFrames based on common columns or indices. It is similar to the SQL join operation.
* It is more powerful and flexible than concatenation when we need to merge DataFrames based on specified key columns.
* Merging is useful when we want to combine information from different DataFrames by matching values in specific columns.

**<code style="background:yellow;color:black">In summary:**</code>

* We use concat() when we want to stack DataFrames along an axis, and they have the same columns or index.
* We use merge() when we want to combine DataFrames based on common columns or indices, especially when we need to perform more complex join operations.

While both methods can be used for some similar tasks, the choice between them depends on the nature of your data and the specific operation you want to perform. merge is more powerful in terms of its ability to handle more complex scenarios, while concat is simpler and more direct for basic stacking of DataFrames.

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">USECASE- Working with IMDB dataset:</code>

**Downloading imdb dataset from google drive link**

In [20]:
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: C:\Users\Lenovo\OneDrive\Desktop\Scaler\Module 5- Python Libraries\My Practice\Class Practice\movies.csv

  0%|          | 0.00/112k [00:00<?, ?B/s]
100%|##########| 112k/112k [00:00<00:00, 620kB/s]
100%|##########| 112k/112k [00:00<00:00, 617kB/s]
Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: C:\Users\Lenovo\OneDrive\Desktop\Scaler\Module 5- Python Libraries\My Practice\Class Practice\directors.csv

  0%|          | 0.00/65.4k [00:00<?, ?B/s]
100%|##########| 65.4k/65.4k [00:00<00:00, 722kB/s]


**Loading the dataset in pandas. We have 2 datsets in imdb dataset**

In [21]:
movies = pd.read_csv("movies.csv")
directors = pd.read_csv("directors.csv")

**Checking the movies dataset** 

In [22]:
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

**We dont want the unnecessary column called "unnamed", so we will remove that permanently**

In [23]:
movies.drop(columns = ["Unnamed: 0"], inplace = True)

movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

**Checking the directors dataset** 

In [24]:
directors.head()

Unnamed: 0      director_name    id gender
0           0      James Cameron  4762   Male
1           1     Gore Verbinski  4763   Male
2           2         Sam Mendes  4764   Male
3           3  Christopher Nolan  4765   Male
4           4     Andrew Stanton  4766   Male

**We dont want the unnecessary column called "unnamed", so we will remove that permanently**

In [25]:
directors.drop(columns = ["Unnamed: 0"], inplace = True)

directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Ques 1:</code> How do we check how many rows and columns are there in the directors dataset?** By using the code below-

In [26]:
directors.shape

(2349, 3)

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Ques 2:</code> How do we check in movies dataset, the number of movies directed by each director_id?** By using the code below-

In [27]:
movies["director_id"].value_counts()

4799    26
4809    19
5087    19
5457    18
4779    16
        ..
5105     5
4849     5
5096     5
5088     5
6204     5
Name: director_id, Length: 199, dtype: int64

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Ques 3:</code> For every the "director_id" in movies dataset, how to find if its present in "id" column of directors table or not?**

### <code style="background:yellow;color:black">isin() method:</code>

* **The <code style="background:yellow;color:black">isin() method</code> is a boolean indexing method that is used to filter data frames based on the values present in a particular column. It returns a boolean Series indicating whether each element in the Series is contained in the specified values.**
* The isin() method in pandas is used to filter data frames. 
* **<code style="background:yellow;color:black">It's similar to a SQL subquery in the sense that it allows us to filter rows based on whether a particular column's values are present in a specified list or another DataFrame.</code>**
* Here in the example below, **True** means that particular director is present in directors table. 

In [28]:
movies["director_id"].isin(directors["id"])

0       True
1       True
2       True
3       True
4       True
        ... 
1460    True
1461    True
1462    True
1463    True
1464    True
Name: director_id, Length: 1465, dtype: bool

**<code style="background:yellow;color:black">Ques 3:</code> Check if all the directors in the movie dataset are present in the director table or not.**

* **In numpy we have learnt a method called np.all(). So, we can use that method here. <code style="background:yellow;color:black">This is an example of numpy and pandas working together.</code>** 
* If np.all(movies["director_id"].isin(directors["id"])) is True, that means every director_id in my movies table is part of my directors table.
* True means they are present and False means anyone of them is not present.

In [29]:
np.all(movies["director_id"].isin(directors["id"]))

True

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">describe() method:</code>

* **In pandas, <code style="background:yellow;color:black">the describe() method</code> is used to generate descriptive statistics of a DataFrame or Series.** 
* **It provides a summary of the central tendency, dispersion, and shape of the distribution of a dataset.**
* When applied to a DataFrame, it returns a DataFrame containing the summary statistics for each numerical column.
* **describe() is a very useful pre-defined method in pandas. It will give us statistical information about every numerical column.** 
* It gives us for every numerical column, the count of the values, average or mean of the values, standard deviation of all the values and so on.  
* **There are 3 text based or string based columns in the dataset. If we want to have some statistical information about the text based columns as well, we have to pass explicitly pass the parameter "include = object" in the describe() method.**
* We note that the statical information about the text based columns is different. We have count, unique, top and frequency for numerical columns. Because maximum, minimum, mean or percentiles of textual or categorical column doesnt make sense, its not included by default.
* Top means which value is occuring the most number of times. In "day" column, "Friday" is occuring the most number of times and that is given by freq which is "654".

In [30]:
movies.describe()

id        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count  director_id         year  
count   1465.000000  1465.000000  1465.000000  
mean    1146.396587  5040.192491  2002.615017  
std     1578.077438   258.059631     8.680141  
min        1.000000  4762.000000  1976.000000  
25%      216.000000  4845.000000  1998.000000  
50%      571.000000  4964.000000  2004.000000  
75%     1387.000000  5179.000000  2009.000000  
max    13752.000000  6204.000000  2016.000000

In [31]:
movies.describe(include = object)

title month     day
count     1465  1465    1465
unique    1465    12       7
top     Avatar   Dec  Friday
freq         1   193     654

**We can check maximum director_id like this also and can cross check with the result above**

In [32]:
movies["director_id"].max()

6204

**We can check minimum diretor_id like this also and can cross check with the result above**

In [33]:
movies["director_id"].min()

4762

In [34]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

<hr style="border: 1px solid gray;">

* **These movies are having very very hugh budgets. Their units should be in millions of dollars but in our dataset, its not in million unit. So, its not that readble, means unnecessarily we have to read these many zeroes. Just to improve the readabilty, we make it in millions of dollars instead of all these zeroes.** 
* **So, we are writing a code to convert the numbers into millions as a unit like this below-** This will apply automatically to every row in that column. 
* **One important thing to note is we can rename the header of the budget column as "budget (in millions)" as well.** This we have illustrated in next code cell.

In [35]:
movies["budget"] = movies["budget"] / 1000000

movies.head()

id  budget  popularity     revenue  \
0  43597   237.0         150  2787965087   
1  43598   300.0         139   961000000   
2  43599   245.0         107   880674609   
3  43600   250.0         112  1084939099   
4  43602   258.0         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [36]:
movies.rename(columns = {"budget" : "budget (in millions)"}, inplace = True)

movies.head()

id  budget (in millions)  popularity     revenue  \
0  43597                 237.0         150  2787965087   
1  43598                 300.0         139   961000000   
2  43599                 245.0         107   880674609   
3  43600                 250.0         112  1084939099   
4  43602                 258.0         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

**Doing the same unit conversion into millions on the revenue column**

In [37]:
movies["revenue"] = movies["revenue"] / 1000000

movies.head()

id  budget (in millions)  popularity      revenue  \
0  43597                 237.0         150  2787.965087   
1  43598                 300.0         139   961.000000   
2  43599                 245.0         107   880.674609   
3  43600                 250.0         112  1084.939099   
4  43602                 258.0         115   890.871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [38]:
movies.rename(columns = {"revenue" : "revenue (in millions)"}, inplace = True)

movies.head()

id  budget (in millions)  popularity  revenue (in millions)  \
0  43597                 237.0         150            2787.965087   
1  43598                 300.0         139             961.000000   
2  43599                 245.0         107             880.674609   
3  43600                 250.0         112            1084.939099   
4  43602                 258.0         115             890.871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday